File: src/jimn/arc.py
"""
arc.
"""
from math import pi
from jimn.elementary_path import ElementaryPath
from jimn.bounding_box import BoundingBox
from jimn.point import Point
from jimn.utils.math import circles_intersections, line_circle_intersections, \
    vline_circle_intersections, compute_arc_centers
from jimn.utils.precision import is_almost
from jimn.displayable import tycat
from jimn.segment import Segment
from jimn.utils.debug import is_module_debugged
from jimn.utils.caching import cached_last_arg


class Arc(ElementaryPath):
    """
    arc class (endpoints, radius, orientation).
    """
    def __init__(self, radius, points, center, reversed_direction=False):
        """
        build an arc out of two endpoints, a radius, a center, a direction.
        it is the user's responsibility to ensure points order is coherent
        with center and direction.
        in doubt, use "compute arc".
        """
        self.radius = radius
        self.center = center
        self.reversed_direction = reversed_direction
        super().__init__(points)
        assert self.radius > 0, "0 or negative radius"

    def copy(self):
        """
        return a deepcopy of given arc.
        """
        return Arc(self.radius, [p.copy() for p in self.endpoints],
                   self.center.copy(), self.reversed_direction)

    def adjust_center(self):
        """
        center was not completely right.
        recompute it
        """
        possible_centers = compute_arc_centers(self.radius, self.endpoints)
        distances = [
            c.distance_to(self.center) for c in possible_centers
        ]
        if distances[0] < distances[1]:
            self.center = possible_centers[0]
        else:
            self.center = possible_centers[1]

    def horizontal_split(self):
        """
        split ourselves on arcs such that for any given x, each arc
        only has one point.
        return array of sub-arcs.
        """
        extreme_points = [self.center + Point([f*self.radius, 0])
                          for f in (-1, 1)]
        for extremum in extreme_points:
            if self.endpoints[0].is_almost(extremum) or\
                    self.endpoints[1].is_almost(extremum):
                continue
            if self.contains(extremum):
                return [
                    Arc(self.radius, [self.endpoints[0], extremum],
                        self.center, self.reversed_direction),
                    Arc(self.radius, [extremum, self.endpoints[1]],
                        self.center, self.reversed_direction)
                ]
        return [self]

    def length(self):
        """
        return arc length.
        """
        return self.radius*self.angle()

    def angle(self):
        """
        return normalized angle of points with center.
        """
        angle = self.center.angle_with(self.endpoints[0]) - \
            self.center.angle_with(self.endpoints[1])
        angle = angle % (2*pi)
        return angle

    def correct_endpoints_order(self):
        """
        when creating arcs in offsetter we need to invert points
        order when distance is more than half of circle
        """
        if self.angle() > pi:
            self.reversed_direction = True
        return self

    def horizontal_intersections_at(self, intersecting_y, xmin, xmax):
        """
        intersections with horizontal line at given y.
        returns array of points.
        """
        segment = Segment([Point([xmin, intersecting_y]),
                           Point([xmax, intersecting_y])])
        intersections = self.intersections_with_segment(segment)
        intersections = [Point([i.get_x(), intersecting_y])
                         for i in intersections]
        return intersections

    def reverse(self):
        """
        return new arc of reversed orientation.
        """
        return Arc(self.radius, list(reversed(self.endpoints)),
                   self.center, not self.reversed_direction)

    def get_bounding_box(self):
        """
        bounding box for arc.
        for now, not tight
        """
        box = BoundingBox.empty_box(2)
        box.add_point(self.center + Point([self.radius, self.radius]))
        box.add_point(self.center - Point([self.radius, self.radius]))
        return box

    def contains(self, point):
        """return true if point is inside arc"""
        if not is_almost(self.center.distance_to(point), self.radius):
            return False
        return self.contains_circle_point(point)

    def intersections_with_arc(self, other):
        """
        intersects with an arc.
        returns up to two points.
        """
        points = circles_intersections(self.center, other.center,
                                       self.radius, other.radius)
        return [
            p for p in points
            if self.contains_circle_point(p) and
            other.contains_circle_point(p)
        ]

    def intersections_with_segment(self, intersecting_segment):
        """
        intersects with a segment.
        returns up to two points.
        """
        points = line_circle_intersections(
            intersecting_segment.endpoints,
            self.center,
            self.radius,
        )
        return [
            p for p in points
            if self.contains_circle_point(p) and
            intersecting_segment.contains(p)
        ]

    def svg_content(self):
        """
        svg for tycat.
        """
        # display first point to know orientation
        arc_string = self.endpoints[0].svg_content()

        # display arc
        x_1, y_1, x_2, y_2 = [
            c for p in self.endpoints
            for c in p.coordinates
        ]
        if self.reversed_direction:
            sweep_flag = 0
        else:
            sweep_flag = 1
        arc_string += self.center.svg_content()

        arc_string += '<path d="M{},{} A{},{} 0 0,{} {},{}" fill="none"/>'.format(
            x_1, y_1,
            self.radius, self.radius,
            sweep_flag,
            x_2, y_2)
        return arc_string

    def path_string(self):
        """
        return svg code for including arc in a svg path.
        """
        end = self.endpoints[1]
        if self.reversed_direction:
            sweep_flag = 0
        else:
            sweep_flag = 1

        return 'A{},{} 0 0,{} {},{}'.format(
            self.radius, self.radius, sweep_flag, *end.coordinates)



    def vertical_intersection_at(self, intersecting_x):
        """return y of lowest intersection given vertical line"""
        min_point, max_point = sorted(self.endpoints)
        if is_almost(min_point.get_x(), intersecting_x):
            return min_point.get_y()
        elif is_almost(max_point.get_x(), intersecting_x):
            return max_point.get_y()

        intersections = \
            vline_circle_intersections(intersecting_x,
                                       self.center, self.radius)

        candidates = [i for i in intersections
                      if self.contains_circle_point(i)]
        if __debug__ and not candidates:
            print(self, intersecting_x, [str(i) for i in intersections])
            tycat(self, intersections)
            raise Exception("no intersections")
        intersecting_ys = [i.get_y() for i in candidates]
        return min(intersecting_ys)

    def translate(self, translation):
        """
        translates arc by a given translation vector
        """
        return Arc(self.radius, [p+translation for p in self.endpoints],
                   self.center + translation, self.reversed_direction)

    def intersections_with(self, other):
        """
        return array of intersections with arc or segment.
        """
        if isinstance(other, Segment):
            intersections = self.intersections_with_segment(other)
        else:
            intersections = self.intersections_with_arc(other)

        if __debug__:
            if is_module_debugged(__name__):
                print("intersections are:")
                tycat(self, other, intersections)
        return intersections

    def contains_circle_point(self, point):
        """
        return true if point (on circle) is inside arc.
        """
        if point.is_almost(self.endpoints[0]) or \
                point.is_almost(self.endpoints[1]):
            return True
        start_angle = self.center.angle_with(self.endpoints[1])
        angles = [
            self.center.angle_with(p) - start_angle
            for p in (point, self.endpoints[0])
        ]
        adjusted_angles = []
        for angle in angles:
            if angle < 0:
                angle += 2*pi
            adjusted_angles.append(angle)
        # assert adjusted_angles[1] <= pi : TODO : fix
        if self.reversed_direction:
            return adjusted_angles[1] < adjusted_angles[0]
        else:
            return adjusted_angles[0] < adjusted_angles[1]

    @cached_last_arg
    def sweeping_key(self, current_x):
        """
        return key used in sweeping line algorithms for comparing paths.
        key is : intersection with vline at current_x, direction angle
        leaving intersection (following tangent), final angle (towards
        destination)
        """
        if __debug__:
            x_coordinates = sorted([p.get_x() for p in self.endpoints])
            if not x_coordinates[0] <= current_x <= x_coordinates[1]:
                print("path is", self, "current x is:", current_x)
                raise Exception("non comparable paths in tree")

        # start by finding the path's y for current x
        point_key = Point([current_x,
                           self.vertical_intersection_at(current_x)])

        if self.endpoints[0] < self.endpoints[1]:
            first_point, last_point = self.endpoints
        else:
            last_point, first_point = self.endpoints

        terminal_angle = (pi/2 - first_point.angle_with(last_point)) % pi

        # we need to round so that segments tangent to some arcs
        # will get exact same angles
        # TODO: why 10 ?
        outgoing_angle = round(
            (pi - self.center.angle_with(point_key)) % pi, 10)
        angles = (outgoing_angle, terminal_angle)

        # now just reverse angles based on direction
        if last_point.is_almost(point_key):
            full_key = (point_key, -angles[0], -angles[1])
        else:
            full_key = (point_key, angles[0], angles[1])
        return full_key

    def clip(self, center, size):
        # TODO: see segment.py
        return self

    def __eq__(self, other):
        if not isinstance(other, Arc):
            return False
        if self.endpoints != other.endpoints:
            return False
        if self.radius != other.radius:
            return False
        if self.reversed_direction != other.reversed_direction:
            return False
        return True

    def __str__(self):
        return "Arc(\n    " + str(self.radius) + ",\n    [\n        " + \
            str(self.endpoints[0]) + ",\n        " + str(self.endpoints[1]) \
            + "\n    ],\n    " + str(self.center) + ",\n    " \
            + str(self.reversed_direction) \
            + "\n)"

    def __hash__(self):
        return hash(tuple(self.endpoints)) ^ hash(self.radius) ^ \
            hash(self.reversed_direction)

    def __lt__(self, other):
        raise Exception("arc comparison still in use")

File: rust/jimn/src/arc.rs
//! provides the `Arc` class.
use std::iter::{empty, once};
use std::f64::consts::{FRAC_PI_2, PI};
use {ElementaryPath, Point, Segment};
use quadrant::{Quadrant, Shape};
use utils::precision::is_almost;
use utils::coordinates_hash::PointsHash;
use utils::ArrayMap;

/// Oriented arc segment.
#[derive(Debug, Clone, Copy, PartialOrd, PartialEq)]
pub struct Arc {
    /// Starting point
    pub start: Point,
    /// Ending point
    pub end: Point,
    /// Center
    pub center: Point,
    /// Radius
    pub radius: f64,
}

impl Arc {
    /// Create a new arc.
    pub fn new(start: Point, end: Point, center: Point, radius: f64) -> Arc {
        let mut arc = Arc {
            start,
            end,
            center,
            radius,
        };
        if !(is_almost(arc.center.distance_to(&arc.start), arc.radius)
            && is_almost(arc.center.distance_to(&arc.end), arc.radius))
        {
            arc.adjust_center();
        }
        arc
    }

    /// Given center was not completely right, move it slightly.
    /// This can happen for example when endpoints have been rounded.
    fn adjust_center(&mut self) {
        self.center = *self.compute_centers()
            .iter()
            .min_by(|c1, c2| {
                c1.distance_to(&self.center)
                    .partial_cmp(&c2.distance_to(&self.center))
                    .unwrap()
            })
            .unwrap();
    }

    /// Return array of the two centers we could have.
    fn compute_centers(&self) -> Vec<Point> {
        // we do some geometry to avoid too complex equations.
        // take start as origin
        let translated_end = self.end - self.start;
        // find bisector
        let middle = translated_end / 2.0;
        let bisector_point = middle + translated_end.perpendicular_vector();
        let line = Segment::new(middle, bisector_point);
        let origin = Point::new(0.0, 0.0);
        let intersections = line_circle_intersections(&line, &origin, self.radius);
        let centers: Vec<_> = intersections.map(|i| self.start + i).collect();
        assert_eq!(centers.len(), 2);
        centers
    }

    /// Return normalized angle of points with center.
    pub fn angle(&self) -> f64 {
        (self.center.angle_with(&self.start) - self.center.angle_with(&self.end) + 2.0 * PI)
            % (2.0 * PI)
    }

    /// Return the arc's length.
    pub fn length(&self) -> f64 {
        let angle = self.angle();
        if angle > PI {
            ((2.0 * PI) - angle) * self.radius
        } else {
            angle * self.radius
        }
    }

    /// Do we contain given point ?
    pub fn contains(&self, point: &Point) -> bool {
        if self.start.is_almost(point) || self.end.is_almost(point) {
            true
        } else if is_almost(self.center.distance_to(point), self.radius) {
            self.contains_circle_point(point)
        } else {
            false
        }
    }

    /// Do we contain given point on circle but not as endpoint ?
    pub fn strictly_contains(&self, point: &Point) -> bool {
        if self.start.is_almost(point) || self.end.is_almost(point) {
            false
        } else {
            self.contains_circle_point(point)
        }
    }

    /// Do we contain given point which is on our circle ?
    pub fn contains_circle_point(&self, point: &Point) -> bool {
        let s = Segment::new(self.start, self.end);
        let s2 = Segment::new(self.center, *point);
        s.intersection_with(&s2).is_some()
    }

    /// Intersect ourselves with horizontal line at given y.
    /// pre-condition: there is exactly one intersection
    ///
    /// # Example
    /// ```
    /// use jimn::{Point, Arc};
    /// let arc = Arc::new(Point::new(1.0, 0.0), Point::new(0.0, -1.0), Point::new(0.0, 0.0), 1.0);
    /// let half_coordinate = 1.0/(2.0 as f64).sqrt();
    /// let half_point = arc.horizontal_line_intersection(-half_coordinate);
    /// assert!(half_point.is_almost(&Point::new(half_coordinate, -half_coordinate)));
    /// ```
    pub fn horizontal_line_intersection(&self, y: f64) -> Point {
        // we use pythagoras
        let side_length = (y - self.center.y).abs();
        if is_almost(side_length, self.radius) {
            return Point::new(self.center.x, y);
        }
        if side_length > self.radius {
            panic!("no arc hline intersection");
        }
        let other_side_length = (self.radius * self.radius - side_length * side_length).sqrt();
        let candidate_point = Point::new(self.center.x - other_side_length, y);
        if self.contains_circle_point(&candidate_point) {
            candidate_point
        } else {
            let candidate_point2 = Point::new(self.center.x + other_side_length, y);
            if self.contains_circle_point(&candidate_point2) {
                candidate_point2
            } else {
                panic!("no arc hline");
            }
        }
    }

    /// Return angle for tangent at given point.
    /// pre-condition: we contain given point.
    ///
    /// # Example
    ///
    /// ```
    /// use jimn::{Point, Arc};
    /// use std::f64::consts::PI;
    /// use jimn::utils::precision::is_almost;
    ///
    /// let arc = Arc::new(Point::new(1.0, 0.0), Point::new(0.0, -1.0), Point::new(0.0, 0.0), 1.0);
    /// let half_coordinate = 1.0/(2.0 as f64).sqrt();
    /// assert!(is_almost(arc.tangent_angle(&Point::new(half_coordinate, -half_coordinate)),
    ///                   PI/4.0));
    /// ```
    pub fn tangent_angle(&self, tangent_point: &Point) -> f64 {
        let base_angle = self.center.angle_with(tangent_point);
        (base_angle + FRAC_PI_2) % PI
    }

    /// Split given `Arc` in possible two so that for any given y, each arc
    /// only has one point.
    /// Does not return anything if arc requires no splitting.
    pub fn split_for_unique_y(&self, rounder: &mut PointsHash) -> Vec<ElementaryPath> {
        for direction in &[1.0f64, -1.0f64] {
            let extremum = self.center + Point::new(0.0, self.radius * *direction);
            if self.strictly_contains(&extremum) {
                let rounded_extremum = rounder.hash_point(&extremum);
                return vec![
                    ElementaryPath::Arc(Arc::new(
                        self.start,
                        rounded_extremum,
                        self.center,
                        self.radius,
                    )),
                    ElementaryPath::Arc(Arc::new(
                        rounded_extremum,
                        self.end,
                        self.center,
                        self.radius,
                    )),
                ];
            }
        }
        vec![ElementaryPath::Arc(*self)]
    }

    /// Iterate on all points obtained when intersecting with given Arc.
    pub fn intersections_with_arc<'a>(
        &'a self,
        other: &'a Self,
    ) -> impl Iterator<Item = Point> + 'a {
        circles_intersections(&self.center, &other.center, self.radius, other.radius)
            .filter(move |p| self.contains_circle_point(p) && other.contains_circle_point(p))
    }

    /// Iterate on all points obtained when intersecting with given Segment.
    pub fn intersections_with_segment<'a>(
        &'a self,
        other: &'a Segment,
    ) -> impl Iterator<Item = Point> + 'a {
        line_circle_intersections(other, &self.center, self.radius)
            .filter(move |p| self.contains_circle_point(p) && other.contains(p))
    }
}

impl Shape for Arc {
    /// Return quadrant for display (also display center).
    /// This quadrant is not tight.
    fn get_quadrant(&self) -> Quadrant {
        Quadrant {
            min_coordinates: vec![self.center.x - self.radius, self.center.y - self.radius],
            max_coordinates: vec![self.center.x + self.radius, self.center.y + self.radius],
        }
    }

    fn svg_string(&self) -> String {
        let center_string = self.center.svg_string();
        // go always for the small arc
        let sweep_flag = if self.angle() > PI { 1 } else { 0 };
        let arc_string = format!(
            "<path d=\"M{},{} A{},{} 0 0,{} {},{}\" fill=\"none\"/>",
            self.start.x,
            self.start.y,
            self.radius,
            self.radius,
            sweep_flag,
            self.end.x,
            self.end.y
        );
        // now draw a small arrow indicating orientation
        let mut middle_angle =
            (self.center.angle_with(&self.start) + self.center.angle_with(&self.end)) / 2.0;
        // we need to figure out where is the middle point between two candidates
        let angles = [middle_angle, middle_angle + PI];
        let possible_points =
            angles.map(|a| self.center + Point::new(a.cos(), a.sin()) * self.radius);
        let distances = possible_points.map(|p| self.start.distance_to(p));
        let tangent_point = if distances[0] < distances[1] {
            possible_points[0]
        } else {
            middle_angle += PI;
            possible_points[1]
        };
        let tangent_angle = if sweep_flag == 1 {
            middle_angle + FRAC_PI_2
        } else {
            middle_angle - FRAC_PI_2
        };
        let arrow_string = format!(
            "<use xlink:href=\"#a\" x=\"{}\" y=\"{}\" transform=\"rotate({} {} {})\"/>",
            tangent_point.x,
            tangent_point.y,
            tangent_angle * 180.0 / PI,
            tangent_point.x,
            tangent_point.y
        );
        center_string + &arc_string + &arrow_string
    }
}

fn line_circle_intersections<'a>(
    segment: &'a Segment,
    center: &'a Point,
    radius: f64,
) -> impl Iterator<Item = Point> + 'a {
    let d = segment.end - segment.start;
    let c = center - segment.start;
    // segment points are at alpha * d
    // distance(alpha * d, center) = r

    // (xc-alpha*xd)**2 + (yc-alpha*yd)**2 - r**2 = 0

    // xc**2 + alpha**2*xd**2 -2*alpha*xc*xd
    // yc**2 + alpha**2*yd**2 -2*alpha*yc*yd
    // - r**2 = 0
    let a = d.x * d.x + d.y * d.y;
    let b = (c.x * d.x + c.y * d.y) * (-2.0);
    let c = c.x * c.x + c.y * c.y - radius * radius;
    let solutions = solve_quadratic_equation(a, b, c);
    solutions.into_iter().map(move |s| segment.start + d * s)
}

fn solve_quadratic_equation(a: f64, b: f64, c: f64) -> Vec<f64> {
    let delta = b * b - a * c * 4.0;
    if is_almost(delta.abs().sqrt(), 0.0) {
        if is_almost(a, 0.0) {
            Vec::new()
        } else {
            vec![-b / (a * 2.0)]
        }
    } else if delta < 0.0 {
        Vec::new()
    } else {
        vec![
            (-b - delta.sqrt()) / (a * 2.0),
            (-b + delta.sqrt()) / (a * 2.0),
        ]
    }
}

fn circles_intersections(c1: &Point, c2: &Point, r1: f64, r2: f64) -> Box<Iterator<Item = Point>> {
    // TODO: should we unbox everything
    // I just solved all equations to end up with this.
    let d = c1.distance_to(c2);
    if is_almost(d, 0.0) {
        Box::new(empty()) // common center
    } else {
        let (x1, y1) = c1.coordinates();
        let (x2, y2) = c2.coordinates();
        let l = if is_almost(r1, r2) {
            d / 2.0
        } else {
            (r1 * r1 - r2 * r2) / (d * 2.0) + d / 2.0
        };

        if is_almost(r1, l) {
            // only one intersection
            Box::new(once(Point::new(
                l / d * (x2 - x1) + x1,
                l / d * (y2 - y1) + y1,
            )))
        } else if (r1 < l) || (r1.abs() < l.abs()) {
            Box::new(empty()) // too far away
        } else {
            let h = (r1 * r1 - l * l).sqrt();
            let p1 = Point::new(
                l / d * (x2 - x1) + h / d * (y2 - y1) + x1,
                l / d * (y2 - y1) - h / d * (x2 - x1) + y1,
            );
            let p2 = Point::new(
                l / d * (x2 - x1) - h / d * (y2 - y1) + x1,
                l / d * (y2 - y1) + h / d * (x2 - x1) + y1,
            );
            if p1.is_almost(&p2) {
                Box::new(once(p1))
            } else {
                Box::new(once(p1).chain(once(p2)))
            }
        }
    }
}
